rfcs/007-background-jobs.md

```diff
@@ -0,0 +1,350 @@
+at a time. But, if the worker process is blocked for too long (for instance, in
+an overload scenario), it may fail to update its status often enough, and the
+activity monitor MAY re-enqueue the job. However, even in such cases, the
+mutual exclusion constraint can be maintained as long as `update(Tx,...)`
+function is called in the same transaction as the type-specific DB writes, and
+its result is checked for the `worker_conflict` return.
+
+# Implementation Details
+
+## Data Model
+
+`("couch_jobs", "data", Type, JobId) = (Sequence, WorkerLockId, JobOpts)`
+`("couch_jobs", "pending", Type, Priority, JobId) = ""`
+`("couch_jobs", "watches", Type) = Sequence`
+`("couch_jobs", "activity_timeout", Type) = ActivityTimeout`
+`("couch_jobs", "activity", Type, Sequence) = JobId`
+
+The `("couch_jobs", "data", Type, JobId)` entry is the `job table` referenced
+in the term definition and throughout the document.
+
+`Type`, `Priority`, and `JobId` MUST be any type supported by the FoundationDB
+tuple layer.
+
+`JobOpts` MUST be a JSON object encoded as a string. It MAY have `"priority"`,
+`"cancel"`, `"resubmit"`, or `"data"` fields.
+
+## Job Lifecyle Implementation
+
+This section describes how the framework implements some of the API functions.
+
+`add()` :
+  * Add the new job to the main jobs table
+  * If a job with the same `JobId` exists, return an error
+  * `WorkerLockId` is set to `null`.
+
+`accept()` :
+ * Generate a unique `WorkerLockId` UUID.
+ * Attempt to dequeue the item from the pending queue, then assign it the
+   `WorkerLockId` in the jobs table.
+ * Create entry in the `"activity"` subspace.
+ * Update sequence in the `"watches"` subspace for that job's type.
+
+`update()`:
+ * Check if `WorkerLockId` matches, otherwise return `worker_conflict`
+ * Check if `"canceled"` field in the `JobOpts` is `true`, if so, return `canceled`
+ * Delete old `"activity"` sequence entry
+ * Update `JobOpts`
+ * Create a new `"activity"` sequence entry and in main job table
+ * Update `"watches"` sequence for that job type
+
+`finish()`:
+ * Check if `WorkerLockId` matches, otherwise returns `worker_conflict`
+ * Check if `Canceled` is `true`, if so, return `canceled`
+ * Delete old `"activity"` sequence entry
+ * If `"resubmit"` field in `JobOpts` is `true`, re-enqueue the job into the pending queue.
+ * Set job table's `WorkerLockId` to `null`
+
+## Activity Monitor Implementation
+
+There is an activity monitor running for each job type. It performs periodic
+scans of the `"activity"` range with sequences less than `"watches"` sequence for
+that type:
+
+  * On first run:
+    * Read the `ActivityTimeout` value and remember the `"watches"` sequence
+  * On subsequent runs:
+    * Call `get_range()` on `"activity"` entries less than the remembered sequence
+    * Read and store the new `"watches"` sequence
+    * If there are any stale entries returned, re-enqueue them
```

Comment:
When a job is considered staled and needs to be re-enqueues we might to need call a job specific function. For indexes, only 1 worker can update an index at one time. So we will most likely have to create a lock in fdb so that no other workers try and update an index that is already being worked on. If the worker then dies, we would need to remove that lock otherwise that index will never be built. When the activity monitor re-enqueues it, it should also either send an event, or call some registered callback for when an indexing job fails with the job information so that their can be some job indexing clean up.

Reply:
Hopefully we can use the `WorkerLockId` and `worker_conflict` semantics. 

I wonder if this will work:

 * If indexing job update their activity on-time, then in the index update transaction, each job checks if `{"resubmit":true}` was set in the `JobOpts`.  If it is, it would call `index_cleanup(...)` and then call `finish(...)`. 

 * However it's possible the index updater job had failed to update its activity often enough so when calling `update(Tx, ...)` and it suddenly gets back `worker_conflict`. Then another worker may already be running the indexing job. Here I guess it depends if indexing cleanup for this stale job, can be performed while another (new) indexing job is already running. As far as the jobs framework is concerned, as long as indexing cleanup doesn't write to or update the "couch_jobs" that would work. But it would depend on the semantics of cleanup and if concurrent cleanup is allowed.

Reply:
Awesome idea, I think we could use `WorkerLockId` to indicate that a specific worker is building the index. Could we then make the JobId the same for for every build or update of a specific index? Then we can use that to know if an index is already being built and whether we need to add another pending job to update an index or just rely on the current job building the index.

Reply:
I think we could probably turn the JobId into a deterministic Id built from ```(Db, DbVersion, ViewGroup, View)```. That might help with matching the indexing job with the view. Then when we try to `add(...)` the job it will come back with a `{error, duplicate_job}` and then we could `resubmit()`.